### PR TITLE
Add initial `faux-mgs` dev tool

### PR DESCRIPTION
**Pull request by jgallagher**
`faux-mgs` is a standalone CLI app to poke SPs. It's primarily intended as a development tool to support hubris SP work, but at the moment I think it makes sense for it to live alongside the real MGS.

### CODE REVIEW

File: gateway/faux-mgs/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+// On success, returns the request ID we sent.
+fn send_request(
+    log: &Logger,
+    socket: &UdpSocket,
+    addr: SocketAddrV6,
+    kind: RequestKind,
+) -> Result<u32> {
+    static REQUEST_ID: AtomicU32 = AtomicU32::new(1);
```
Comment (andrewjstone) [inline gateway/faux-mgs/src/main.rs:147]: Why is this atomic? For mutability? I assume you made it static so you wouldn't have to pass it as a parameter.

File: gateway/faux-mgs/src/usart.rs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+            match result {
+                Ok(ResponseKind::SerialConsoleWriteAck) => {
+                    if Some(response_id) == request_id {
+                        debug!(log, "received expected ack");
+                        return true;
+                    } else {
+                        debug!(log, "received unexpected ack");
+                    }
+                }
+                Ok(other) => {
```
Comment (andrewjstone) [inline gateway/faux-mgs/src/usart.rs:146]: This seems to imply that we are only handling responses from the SP. Does the SP ever send async messages?

**Review by andrewjstone — APPROVED** [review]
This looks good to me. It's pretty straightforward, barring all the termios chicanery. 
I'm curious why you ended up using mio directly and not tokio, but not enough to hold up merging a dev tool.

File: gateway/faux-mgs/src/main.rs (hunk 1 above)
Comment (jgallagher) [reply gateway/faux-mgs/src/main.rs:147]: Yes, just for simplicity - if this function sends all packets, this allows it to keep its own monotonically increasing counter, freeing the caller up from worrying about it.

File: gateway/faux-mgs/src/usart.rs (hunk 2 above)
Comment (jgallagher) [reply gateway/faux-mgs/src/usart.rs:146]: It definitely can, and MGS proper has a lot more machinery to handle that. `faux-mgs` may eventually need to handle this better, but for now the only async messages the SP can send contain serial console data, which it won't send unless an MGS instance has already attached by sending it data first.

**Comment by jgallagher** [comment]
Yeahhh, I should probably go back and replace `mio` with `tokio`. I was ~90% done before I realized I needed timeouts when reading from stdin, and grabbed mio as a smaller change to implement that. I'll go ahead and merge for now but plan on updating that next time I'm in here making changes.

File: gateway/faux-mgs/src/usart.rs (hunk 2 above)
Comment (andrewjstone) [reply gateway/faux-mgs/src/usart.rs:146]: Thanks. Makes sense.

**Comment by andrewjstone** [comment]
> Yeahhh, I should probably go back and replace `mio` with `tokio`. I was ~90% done before I realized I needed timeouts when reading from stdin, and grabbed mio as a smaller change to implement that. I'll go ahead and merge for now but plan on updating that next time I'm in here making changes.
> 
> Thanks for the review!

Honestly, I don't think it's urgent at all to convert. If you have time and feel like it feel free, but I'm certainly not going to complain if we use mio forever on this.

**Pull request merged**